contracts/src/registry/PermissionedRegistry.sol

```diff
@@ -2,55 +2,264 @@
 pragma solidity >=0.8.13;
 
 import {AccessControl} from "@openzeppelin/contracts/access/AccessControl.sol";
-
+import {IERC165} from "@openzeppelin/contracts/utils/introspection/IERC165.sol";
 import {ERC1155Singleton} from "./ERC1155Singleton.sol";
 import {IERC1155Singleton} from "./IERC1155Singleton.sol";
 import {IRegistry} from "./IRegistry.sol";
 import {IRegistryDatastore} from "./IRegistryDatastore.sol";
-import {IRegistryMetadata} from "./IRegistryMetadata.sol";
 import {BaseRegistry} from "./BaseRegistry.sol";
+import {EnhancedAccessControl} from "./EnhancedAccessControl.sol";
+import {MetadataMixin} from "./MetadataMixin.sol";
+import {RegistryMetadata} from "./RegistryMetadata.sol";
+import {SimpleRegistryMetadata} from "./SimpleRegistryMetadata.sol";
+import {NameUtils} from "../utils/NameUtils.sol";
+import {IPermissionedRegistry} from "./IPermissionedRegistry.sol";
+
+contract PermissionedRegistry is IPermissionedRegistry, BaseRegistry, EnhancedAccessControl, MetadataMixin {
+    mapping(uint256 => address) public tokenObservers;
+
+    uint256 public constant ROLE_REGISTRAR = 1 << 0;
+    uint256 public constant ROLE_REGISTRAR_ADMIN = ROLE_REGISTRAR << 128;
+
+    uint256 public constant ROLE_RENEW = 1 << 1;
+    uint256 public constant ROLE_RENEW_ADMIN = ROLE_RENEW << 128;
+
+    uint256 public constant ROLE_SET_SUBREGISTRY = 1 << 2;
+    uint256 public constant ROLE_SET_SUBREGISTRY_ADMIN = ROLE_SET_SUBREGISTRY << 128;
 
-abstract contract PermissionedRegistry is BaseRegistry {
-    uint96 public constant FLAGS_MASK = 0x7;
-    uint96 public constant FLAG_SUBREGISTRY_LOCKED = 0x1;
-    uint96 public constant FLAG_RESOLVER_LOCKED = 0x2;
-    uint96 public constant FLAG_FLAGS_LOCKED = 0x4;
+    uint256 public constant ROLE_SET_RESOLVER = 1 << 3;
+    uint256 public constant ROLE_SET_RESOLVER_ADMIN = ROLE_SET_RESOLVER << 128;
+
+    uint256 public constant ROLE_SET_FLAGS = 1 << 4;
+    uint256 public constant ROLE_SET_FLAGS_ADMIN = ROLE_SET_FLAGS << 128;
+
+    uint256 public constant MAX_EXPIRY = type(uint64).max;
+    uint96 public constant FLAGS_MASK = 0xffffffff; // 32 bits
```

Comment:
I'm a bit concerned about the bytecode and runtime overhead of all these public constants. What do you think about making them private?

Reply:
That's fine, but we then have to accept front-end brittleness, i.e things could go wrong.